aws/resource_aws_dms_event_subscription.go

```diff
@@ -0,0 +1,251 @@
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		MinTimeout: 10 * time.Second,
+		Delay:      10 * time.Second,
+	}
+
+	// Wait, catching any errors
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("error waiting for DMS event subscription (%s) creation: %s", d.Id(), err)
+	}
+
+	return resourceAwsDmsEventSubscriptionUpdate(d, meta)
+}
+
+func resourceAwsDmsEventSubscriptionUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dmsconn
+
+	request := &dms.ModifyEventSubscriptionInput{
+		Enabled:          aws.Bool(d.Get("enabled").(bool)),
+		SnsTopicArn:      aws.String(d.Get("sns_topic_arn").(string)),
+		SubscriptionName: aws.String(d.Get("name").(string)),
+		SourceType:       aws.String(d.Get("source_type").(string)),
+	}
+
+	if v, ok := d.GetOk("event_categories"); ok {
+		request.EventCategories = expandStringList(v.(*schema.Set).List())
+	}
+
+	log.Println("[DEBUG] DMS update event subscription:", request)
+
+	_, err := conn.ModifyEventSubscription(request)
+
+	if err != nil {
+		return fmt.Errorf("Error updating DMS event subscription: %s", err)
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"modifying"},
+		Target:     []string{"active"},
+		Refresh:    resourceAwsDmsEventSubscriptionStateRefreshFunc(conn, d.Id()),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		MinTimeout: 10 * time.Second,
+		Delay:      10 * time.Second, // Wait 30 secs before starting
+	}
+
+	// Wait, catching any errors
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("error waiting for DMS event subscription (%s) modification: %s", d.Id(), err)
+	}
+
+	return resourceAwsDmsEventSubscriptionRead(d, meta)
+}
+
+func resourceAwsDmsEventSubscriptionRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).dmsconn
+
+	request := &dms.DescribeEventSubscriptionsInput{
+		SubscriptionName: aws.String(d.Id()),
+	}
+
+	log.Println("[DEBUG] DMS read event subscription:", request)
+
+	response, err := conn.DescribeEventSubscriptions(request)
+
+	if isAWSErr(err, dms.ErrCodeResourceNotFoundFault, "") {
+		log.Printf("[WARN] DMS event subscription (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error reading DMS event subscription: %s", err)
+	}
+
+	if response == nil || response.EventSubscriptionsList == nil || len(response.EventSubscriptionsList) == 0 {
+		log.Printf("[WARN] DMS event subscription (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	subscription := response.EventSubscriptionsList[0]
+
+	d.Set("enabled", subscription.Enabled)
+	d.Set("sns_topic_arn", subscription.SnsTopicArn)
+	d.Set("source_type", subscription.SourceType)
+	d.Set("name", d.Id())
+	d.Set("event_categories", flattenStringList(subscription.EventCategoriesList))
+	d.Set("source_ids", flattenStringList(subscription.SourceIdsList))
+
```

Comment:
Tags should be refreshed from the API into the Terraform state for drift detection. Since the DescribeEventSubscriptions API response does not include them, we can use the `keyvaluetags` package functionality to fetch them. To do so though, we need the ARN of the event subscription. The [DMS User Guide section on ARNs](https://docs.aws.amazon.com/dms/latest/userguide/CHAP_Introduction.AWS.ARN.html) does not explicitly list the event subscription ARN format, but I would try `es:` as the resource prefix since that is what [RDS uses](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_Tagging.ARN.html).

In the schema, add:

```go
"arn": {
	Type:     schema.TypeString,
	Computed: true,
},
```

Then here:

```suggestion
	arn := arn.ARN{
		Partition: meta.(*AWSClient).partition,
		Service:   "dms",
		Region:    meta.(*AWSClient).region,
		AccountID: meta.(*AWSClient).accountid,
		Resource:  fmt.Sprintf("es:%s", d.Id()),
	}.String()
	d.Set("arn", arn)

	tags, err := keyvaluetags.DatabasemigrationserviceListTags(conn, arn)

	if err != nil {
		return fmt.Errorf("error listing tags for DMS Event Subscription (%s): %s", arn, err)
	}

	if err := d.Set("tags", tags.IgnoreAws().Map()); err != nil {
		return fmt.Errorf("error setting tags: %s", err)
	}
```